### ISSUE TITLE
Login() takes Credentials instead of UserObjectType

### ISSUE THREAD

**Issue by sproott** [jkettmann/graphql-passport#41]
In types.ts, the login function is declared to accept Credentials, but in buildContext.ts it actually accepts UserObjectType. Is that intentional?

**Comment by gforge** [comment]
Hmm... that seems like an error from my part. The login function is controlled by the main passport library as I understand it: https://github.com/jaredhanson/passport/blob/master/lib/http/request.js#L10-L57

It suggests that the login should accept a full user object. Although - every time I look into the passport code I seem to end up in an evil loop searching for where the different functions are defined and I can't claim that I fully grasp passport. My grasp of `this` is just too weak.

**Comment by sproott** [comment]
Well, I ended up dissecting this library and tweaking it heavily so that I can use it with different credentials than just username and password. If you want to implement this, just make the authenticate method accept an object of generic type Credentials instead of username and password. I actually don't know why this wasn't already done.
https://github.com/jkettmann/graphql-passport/blob/master/src/GraphQLLocalStrategy.ts#L45

**Comment by firmanjml** [comment]
@jkettmann @gforge @sproott 

Hi, is there any solution for this matter? I have been getting mistype errors with SharedPassportContext not providing the correct types.

**Comment by sproott** [comment]
I can make a PR in under a week from now when I get to it.

**Comment by firmanjml** [comment]
> I can make a PR in under a week from now when I get to it. 

This is my first time using passportjs, and i don’t know if i do it right, i overwrite the login type to accept MyUser type as the argument instead of credentials because i do not want to store password in my session. 

Was that intentional? 
Because i thought of exposing id, username, email and roles on session context.

**Comment by sproott** [comment]
Well I assume you don't store password as plaintext anyways. I have only userId in the session and load the whole user object into the context. Then you can pull anything out of it.

**Comment by firmanjml** [comment]
> Well I assume you don't store password as plaintext anyways. I have only userId in the session and load the whole user object into the context. Then you can pull anything out of it. 

What data do you pass in the context.login() argument?

Because the login() argument in my project takes in Credential Type and authenticate() returns UserObject

**Comment by sproott** [comment]
@firmanjml @gforge @jkettmann please have a look at this, it was made rather quickly without testing. But it should make things much easier to understand and use. 

https://github.com/jkettmann/graphql-passport/pull/56

**Comment by GioAc96** [comment]
> In types.ts, the login function is declared to accept Credentials, but in buildContext.ts it actually accepts UserObjectType. Is that intentional?

I suspect that the issue derives from the fact that the user object usually contains two fields that directly correspond to the credentials, which makes it so that most users of this library won't run into this problem, but this is definitely a bug. I'm currently using //@ts-ignore to override the type check complaints, but this needs to be fixed

**Comment by gforge** [comment]
Credentials is only what is needed for the login functionality, e.g. username and password. The context should contain the full user object, I'm not sure why you need a @ts-ignore in your case.

I've been experimenting with a setup that also uses subscriptions with authentication in a bare-bone-test-repo that is a little more detailed than the current documentation (that also needs some minor updates): https://github.com/gforge/subscription_example

**Comment by GioAc96** [comment]
> Credentials is only what is needed for the login functionality, e.g. username and password. The context should contain the full user object, I'm not sure why you need a @ts-ignore in your case.
> 
> I've been experimenting with a setup that also uses subscriptions with authentication in a bare-bone-test-repo that is a little more detailed than the current documentation (that also needs some minor updates): https://github.com/gforge/subscription_example

I use `//@ts-ignore` since my `User` type is not an extension of my `Credentials` type, and therefore I'm not able to pass the User object to the `authenticate()` method. I don't understand why the `login()` method would take credentials that have already been verified by the `authenticate()` method, rathen than a full `User` object that has already been fetched, since ultimately the framework serializes the user and not the credentials.

**Comment by gforge** [comment]
Yeah, it is a typo from my part. The PR #58 should handle that. I also noted that the context actually doesn't have `user` in it but it is under `req.user` (probably best to user `context.getUser()`. @jkettmann mentioned that he was a little busy but hopefully there will be some updates that fix these issues soon.
